---
title: "Volcano Plots Per Cell Type for sc-RNA-Seq Analysis in 10X Genomics data"
author: "Antonia Chroni for SJCRH DNB_BINF_Core"
papersize: a4
fontsize: 11pt
links-as-notes: true
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    df_print: paged
    code_folding: hide
    toc_depth: 2
    highlight: tango
    number_sections: TRUE
  pdf_document:
    toc: TRUE
    highlight: tango
    number_sections: TRUE
    latex_engine: lualatex
    keep_tex: FALSE
    fig_caption: yes
    fig_crop: no
    fig_height: 2
    fig_width: 3
    toc_depth: 2
always_allow_html: TRUE
urlcolor: blue
linkcolor: black
citecolor: blue
geometry: margin=1in
header-includes: 
  - \usepackage{titling}
  - \usepackage{fancyhdr}
  - \usepackage{graphicx}
  - \usepackage{float}
params:
  resolution_list: '.'
  n_value: '.'
  condition_value1: '.'
  condition_value2: '.'
  condition_value3: '.'
  root_dir: './'
  PROJECT_NAME: '.'
  PI_NAME: '.'
  TASK_ID: '.'
  PROJECT_LEAD_NAME: '.'
  DEPARTMENT: '.'
  LEAD_ANALYSTS: '.'
  GROUP_LEAD: '.'
  CONTACT_EMAIL: '.'
  PIPELINE: '.'
  START_DATE: '.'
  COMPLETION_DATE: '.'
---

```{r logo-file, echo=FALSE}
attach(params)
# Insert logo on the top of the html report 
logo_file <- file.path(root_dir, "figures", "img", "DNB-BINF-Core-logo.png")
htmltools::img(src = knitr::image_uri(logo_file), alt = "logo", style = "position:absolute; top:0; left:0; padding:0px; height:120px;")
detach(params)
```

\addtolength{\headheight}{2.0cm} 
\fancypagestyle{plain}{} 
\thispagestyle{fancy}
\fancyhead[L]{\includegraphics[height=120px]{`r logo_file`}}
\renewcommand{\headrulewidth}{0pt}

<style type="text/css">
:root {--DNB_BINF_Core_color: #00427B;}

h1.title {margin-top: 130px;
          margin-bottom: 25px;
          font-size: 36px;}

.nobullet li {list-style-type: none;}

.reporthead {font-size: 20px;}

body { /* Normal */
  font-size: 16px;
  font-style: Arial, Helvetica, sans-serif;}

h1 {color: var(--DNB_BINF_Core_color);
    font-size: 28px;
    margin-top: 50px;}

h2 {color: var(--DNB_BINF_Core_color);
    font-size: 20px;}

.list-group-item.active, .list-group-item.active:focus, .list-group-item.active:hover {
  background-color: var(--DNB_BINF_Core_color);}
</style>

<a href="https://wiki.stjude.org/display/CAB">

</a>

\pagebreak

<div class="reporthead"><br/>
**PI: `r params$PI_NAME`**  
**Project: `r params$PROJECT_NAME`**  
Task: `r params$TASK_ID`  
Project Lead(s): `r params$PROJECT_LEAD_NAME`  
Department: `r params$DEPARTMENT`  

<br />  

DNB Bioinformatics Core Analysis Team: 
<br />  

>**Lead Analyst(s): `r params$LEAD_ANALYSTS`**  
>Group Lead: `r params$GROUP_LEAD`  
<br />
>**Contact E-mail:** `r params$CONTACT_EMAIL`  
>**DNB Bioinformatics Core Pipeline:** `r params$PIPELINE`  

Date started: `r params$START_DATE`  
Date completed:  `r params$COMPLETION_DATE`  
Report generated: `r format(Sys.time(), '%H:%M:%S %Z %m/%d/%Y')` \

Reviewed by: _____________________   Date: ____________ \
</div>
\pagebreak
  
# Information about this notebook

This notebook generates volcano plots to visualize and summarize DE patterns.

A volcano plot visualizes two key metrics for every gene:

- logâ‚‚ fold change (log2FC) â€” the magnitude and direction of expression change
- statistical significance â€” usually represented as â€“logâ‚â‚€(p-value)

By combining these two axes, volcano plots highlight genes that show both large expression changes and strong statistical support, making it easier to identify potentially biologically meaningful upâ€‘ and downâ€‘regulated genes within each cell type.

## ðŸ” Interpretation Guidelines

**Up-regulated genes (red-right side)**

Genes plotted in red on the right side can represent transcriptional programs increased or activated in the given cell type or the condition for the given cell type.

- log2FC > 1 (higher expression in the given cell type or the condition within the given cell type, e.g. Treatment vs Control)
- p-value â‰¤ Î± (statistically significant)

Depending on the DE comparison, these genes may indicate:

- cell-type-specific markers
- pathway activation
- treatment-induced responses
- cell-state transitions
- altered signaling or regulatory activity

**Down-regulated genes (red-left side)**

Genes plotted in red on the left side can represent transcriptional programs decreased or inactivated in the given cell type or the condition for the given cell type.

- log2FC < -1 (lower expression in the given cell type or the condition within the given cell type, e.g. Treatment vs Control)
- p-value â‰¤ Î±

Depending on the DE comparison, these genes may indicate:

- cell-type-specific markers
- suppressed pathways
- inhibited regulatory programs
- decreased cell-typeâ€“specific functions
- loss of activity due to treatment or condition

**Non-significant genes (grey)**

Genes shown in grey are:

- not significant at the chosen p-value threshold (p-value > Î±), and/or
- show only minor fold-change that does not meet the log2FC threshold.

These genes essentially represent background variation or small changes that are unlikely to be biologically meaningful under current criteria.


## ðŸ“Œ Summary

A volcano plot provides a fast and intuitive overview of differentialâ€‘expression (DE) patterns within each cell type and condition.
Note that by default the `EnhancedVolcano()` function uses a cutoff of `1e-05` for Î±.

### Color Guide

| **Color** | **Meaning**              | **Criteria**                                       |   **Volcano side**                                              |
|-----------|--------------------------|----------------------------------------------------|-----------------------------------------------------------------|
| **Red**   | Upâ€‘regulated             | log2FC > 1 & p-value â‰¤ Î±                  |   Right side (positive logâ‚‚FC)                                  |
| **Red**   | Downâ€‘regulated           | log2FC < -1 & p-value â‰¤ Î±                  |   Left side (negative logâ‚‚FC)                                   |
| **Blue**  | Significant p-value only | pâ€‘value â‰¤ Î± but logâ‚‚FC < threshold        |   logâ‚‚FC; Both sides (near 0 logâ‚‚FC)                            |
| **Green** | High foldâ€‘change only    | logâ‚‚FC â‰¥ threshold but pâ€‘value > Î±        |   Left = large negative FC; Right = large positive FC           |
| **Grey**  | Not significant          | p-value > Î± and logâ‚‚FC < threshold        |   Center region around logâ‚‚FC â‰ˆ 0                               |

This visualization helps identify genes that show both statistical significance at the given Î± and potentially biologically meaningful foldâ€‘change in a cellâ€‘type-specific manner.


# Set up

```{r load-library, echo=TRUE}
attach(params)
suppressPackageStartupMessages({
  library(tidyverse)
  library(knitr)
  library(DT)
  library(glue)
  library(EnhancedVolcano)
}) 
```

# Directories and paths to file Inputs/Outputs

```{r set-dir-and-file-names, echo=TRUE}
analysis_dir <- file.path(root_dir, "analyses", "de-go-analysis") 
module_results_dir <- file.path(analysis_dir, "results")
module_plots_dir <- file.path(analysis_dir, "plots")
markers_dir_module <- file.path(analysis_dir, "results", "01-de-analysis")

# Input files
markers_file_overall <- file.path(markers_dir_module, glue::glue("Res_{resolution_list}_Markers_all.tsv"))
markers_file_condition_value1 <- file.path(markers_dir_module, glue::glue("Res_{resolution_list}_Markers_all_condition_value1.tsv"))
markers_file_condition_value2 <- file.path(markers_dir_module, glue::glue("Res_{resolution_list}_Markers_all_condition_value2.tsv"))
markers_file_condition_value3 <- file.path(markers_dir_module, glue::glue("Res_{resolution_list}_Markers_all_condition_value3.tsv"))

# Create plots directory
plots_dir <- file.path(module_plots_dir, "02-volcano-plots") 
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)}

plots_dir_overall <- file.path(plots_dir, "01-volcano-plots-overall") 
if (!dir.exists(plots_dir_overall)) {
  dir.create(plots_dir_overall)}

plots_dir_condition1 <- file.path(plots_dir, "02-volcano-plots-condition1") 
if (!dir.exists(plots_dir_condition1)) {
  dir.create(plots_dir_condition1)}

# Source functions
source(paste0(analysis_dir, "/util/function-create-volcano-plot.R"))
```

```{r echo=FALSE, warning=FALSE}
opts_chunk$set(fig.align='center',
               external=TRUE,
               echo=FALSE,
               warning=FALSE,
               fig.pos='H')
a4width <- 8.3
a4height <- 11.7
```

# Volcano Plots Per Cell Type

The analysis labels the top `r n_value` genes ranked by significance to highlight the most potentially biologically relevant upâ€‘ and downâ€‘regulated features.

## Overall

```{r volcano-plot, fig.width = 8, fig.height = 6, fig.fullwidth = TRUE, echo=TRUE}
all.markers_overall <- read.csv(markers_file_overall, sep = "\t", header = TRUE)
volcano_plot_results <- create_volcano_plot(all.markers = all.markers_overall, n_value = n_value, plots_dir = plots_dir_overall)
```

## Per `r condition_value1`

```{r volcano-plot-condition1, fig.width = 8, fig.height = 6, fig.fullwidth = TRUE, echo=TRUE}
all.markers_condition_value1 <- read.csv(markers_file_condition_value1, sep = "\t", header = TRUE)
volcano_plot_results <- create_volcano_plot(all.markers = all.markers_condition_value1, n_value = n_value, plots_dir = plots_dir_condition1)
```

## Per `r condition_value2`

```{r volcano-plot-condition2, fig.width = 8, fig.height = 6, fig.fullwidth = TRUE, echo=TRUE}
if (!is.null(condition_value2)) {
  
  all.markers_condition_value2 <- read.csv(markers_file_condition_value2, sep = "\t", header = TRUE)

  plots_dir_condition2 <- file.path(plots_dir, "03-volcano-plots-condition2") 
  if (!dir.exists(plots_dir_condition2)) {
    dir.create(plots_dir_condition2)}
  
  volcano_plot_results <- create_volcano_plot(all.markers = all.markers_condition_value2, n_value = n_value, plots_dir = plots_dir_condition2)
}
```

## Per `r condition_value3`

```{r volcano-plot-condition3, fig.width = 8, fig.height = 6, fig.fullwidth = TRUE, echo=TRUE}
if (!is.null(condition_value3)) {
    
  all.markers_condition_value3 <- read.csv(markers_file_condition_value3, sep = "\t", header = TRUE)

  plots_dir_condition3 <- file.path(plots_dir, "04-volcano-plots-condition3") 
  if (!dir.exists(plots_dir_condition3)) {
    dir.create(plots_dir_condition3)}
  
  volcano_plot_results <- create_volcano_plot(all.markers = all.markers_condition_value3, n_value = n_value, plots_dir = plots_dir_condition3)
}
```

```{r echo=FALSE}
detach(params)
```

\pagebreak

# Session Info

```{r echo=FALSE}
sessionInfo()
```

